### PR TITLE
Use BOOST_ASIO_SEPARATE_COMPILATION for Boost.Asio

### LINKED ISSUE #166: Use BOOST_ASIO_SEPARATE_COMPILATION for Boost.Asio
Hi, I just wanted to suggest using `BOOST_ASIO_SEPARATE_COMPILATION` for Boost.Asio.

Most other heavyweight Boost libraries already require separate compilation, but Asio has left its default as header-only for convenience (so users don't have to deal with the complexity of compiling and linking to Boost). But given Bazel already takes care of that complexity, it seems better to compile the library separately and link to it.

To do this, you can just [follow the instructions here](https://www.boost.org/doc/libs/1_72_0/doc/html/boost_asio/using.html#boost_asio.using.optional_separate_compilation), e.g.:

	boost_library(
	    name = "asio",
	    srcs = [
	        "boost/asio/impl/src.cpp",
	        "boost/asio/impl/src.hpp",
	    ],
	    defines = [
	        "BOOST_ASIO_SEPARATE_COMPILATION",
	    ] + select({
	    ...

(Also see the note about `#include <boost/asio/ssl/impl/src.hpp>` for SSL use. I'm not sure if that's best included in the same target or a separate one, if any.)

Thanks!

### PR DESCRIPTION
**Pull request by mehrdadn**
Resolves #166.

### CODE REVIEW

**Comment by nelhage** [comment]
Looks good! Looks like Circle is down for scheduled maintenance right now, but I'll merge once they're back and can run the job. Thanks for your contribution.

**Comment by mehrdadn** [comment]
Thank you!

File: BUILD.boost [hunk 1]
```diff
@@ -323,7 +323,13 @@ boost_library(
 
 boost_library(
     name = "asio",
-    defines = select({
+    srcs = [
+        "boost/asio/impl/src.cpp",
+        "boost/asio/impl/src.hpp",
```
Comment (sehe) [inline BUILD.boost:328]: Sorry for very late question. Trying to understand this rule, what does the srcs array mean? In particular, any reason whey the same source appears twice?

Comment (mehrdadn) [reply BUILD.boost:328]: It specifies the inputs to this rule. Note that it's not the same file; one is `.cpp` and one is `.hpp`.

Comment (sehe) [reply BUILD.boost:328]: Oh my. I looked it over several times, and even followed the documentation link, somehow still missed that. Time for break. Thanks for the reply!

**Pull request merged**